### PR TITLE
Trying to fix issue #63788: "NetworkChange.NetworkAddressChanged event leaks threads on Linux

### LINKED ISSUE #63788: `NetworkChange.NetworkAddressChanged` event leaks threads on Linux
### Description

When you enable HTTP/3 support on Linux and repeatedly get HTTP/2 or HTTP3 web page, ".NET Network Address Change" threads will start to accumulate.

It seems network address change event is triggered from https://source.dot.net/#System.Net.Http/System/Net/Http/SocketsHttpHandler/HttpConnectionPoolManager.cs with call _poolManager.StartMonitoringNetworkChanges();

NetworkAddressChange.Unix.cs will start creating threads with name ".NET Network Address Change":
https://source.dot.net/#System.Net.NetworkInformation/System/Net/NetworkInformation/NetworkAddressChange.Unix.cs

### Reproduction Steps

----------------------------------------------------------------------------
.NET 6 console project, source code example:
----------------------------------------------------------------------------

```C#
using System.Diagnostics;

static async Task ThreadDiagnostics()
{
    var pid = Process.GetCurrentProcess().Id;
    var psi = new ProcessStartInfo("ps", $"-T -p {pid}");
    psi.RedirectStandardOutput = true;

    using var process = Process.Start(psi);
    if (process == null)
    {
        throw new Exception("Could not create process");
    }

    int threadCountNetworkAddressChange = 0;
    var output = await process.StandardOutput.ReadToEndAsync();
    using (var sr = new StringReader(output))
    {
        while (true)
        {
            var line = await sr.ReadLineAsync();
            if (line == null)
            {
                break;
            }

            if (line.IndexOf(".NET Network Ad") > 0)
            {
                threadCountNetworkAddressChange++;
            }

            //NOTE:
            //we are searching for threads containing ".NET Network Ad"

            //new Thread(s => LoopReadSocket((int)s!))
            //{
            //    IsBackground = true,
            //    Name = ".NET Network Address Change"
            //}.UnsafeStart(newSocket);
            //
            //https://source.dot.net/#System.Net.NetworkInformation/System/Net/NetworkInformation/NetworkAddressChange.Unix.cs
        }
    }

    //Console.WriteLine(output);
    Console.WriteLine($"{DateTime.Now} '.NET Network Address Change' threads: {threadCountNetworkAddressChange}");
}

for (var i = 0; i < 60; i++)
{
    try
    {
        using (var client = new HttpClient())
        {
            client.DefaultVersionPolicy = HttpVersionPolicy.RequestVersionExact;
            using var req = new HttpRequestMessage(HttpMethod.Get, "https://cloudflare-quic.com/"); //HTTP/2 or HTTP/3 enabled web server
            req.Version = System.Net.HttpVersion.Version20;

            using var res = await client.SendAsync(req);
            res.EnsureSuccessStatusCode();
        }

        await ThreadDiagnostics();
    }
    catch (Exception ex)
    {
        Console.WriteLine($"Unhandled exception: {ex}");
    }

    await Task.Delay(1000);
}
```

----------------------------------------------------------------------------
Dockerfile with installed libmsquic and required tools (procps, curl)
----------------------------------------------------------------------------

- See https://aka.ms/containerfastmode to understand how Visual Studio uses this Dockerfile to build your images for faster debugging.

```
FROM mcr.microsoft.com/dotnet/runtime:6.0-bullseye-slim AS base
WORKDIR /app

#install some tools
RUN apt-get update && apt-get install -y procps curl

#install libmsquic
#https://docs.microsoft.com/en-us/dotnet/core/extensions/httpclient-http3
RUN curl https://packages.microsoft.com/debian/11/prod/pool/main/libm/libmsquic/libmsquic_1.9.0_amd64.deb --output libmsquic.deb
RUN dpkg -i libmsquic.deb

FROM mcr.microsoft.com/dotnet/sdk:6.0 AS build
WORKDIR /src
COPY ["HttpClientMemoryLeak/HttpClientMemoryLeak.csproj", "HttpClientMemoryLeak/"]
RUN dotnet restore "HttpClientMemoryLeak/HttpClientMemoryLeak.csproj"
COPY . .
WORKDIR "/src/HttpClientMemoryLeak"
RUN dotnet build "HttpClientMemoryLeak.csproj" -c Release -o /app/build

FROM build AS publish
RUN dotnet publish "HttpClientMemoryLeak.csproj" -c Release -o /app/publish

FROM base AS final
WORKDIR /app
COPY --from=publish /app/publish .
ENTRYPOINT ["dotnet", "HttpClientMemoryLeak.dll"]
```

### Expected behavior

".NET Network Address Change" should not accumulate, especially if HTTP connections are not pooled.

Check https://source.dot.net/#System.Net.Http/System/Net/Http/SocketsHttpHandler/HttpConnectionPoolManager.cs

```
bool avoidStoringConnections =
                settings._maxConnectionsPerServer == int.MaxValue &&
                (settings._pooledConnectionIdleTimeout == TimeSpan.Zero ||
                 settings._pooledConnectionLifetime == TimeSpan.Zero);
```

### Actual behavior

".NET Network Address Change" threads will start to accumulate and consume resources / memory.

### Regression?

_No response_

### Known Workarounds

_No response_

### Configuration

_No response_

### Other information

_No response_

### PR DESCRIPTION
**Pull request by vdjuric**
Adding "s_socketCreateContext" counter which changes each time CreateSocket() method is called. Method LoopReadSocket can then use this information to exit thread.

Fixes #63788

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=63963) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=63963) before we can accept your contribution.<br/><br/>:x: vdjuric [sign now](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=63963)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/runtime?pullRequest=63963) it.</sub>

**Comment by ManickaP** [comment]
Do you think you could also bring the repro code as a test?

**Comment by vdjuric** [comment]
Hi,

Yes, I will try to add minimal functional test for this issue, but first I have to setup local Linux development environment in WSL, so I can rebuild .NET Core locally from fork and run unit tests. Might take some time :)

**Comment by vdjuric** [comment]
Hi,

After creating initial commit (https://github.com/vdjuric/runtime-NetworkChange-unix-thread-leak/commit/5c33cce47264d85ada6b9a6921ac210d76d776d0) and then writing function test, I noticed issue with leaking threads was still not fixed. I investigated a little bit and it seems native code in pal_networkchange.c, method SystemNative_ReadEvents was also blocking thread because recvmsg waited indefinitely...

I then added timeout for receive method (currently set to one second) and it seems this fixed the issue. Functional test now reports 0 threads after all event subscribers were removed.

**NOTE: I am not Linux and not C expert, so I suggest good review on this one :)**

File: src/native/libs/System.Native/pal_networkchange.c [hunk 1]
```diff
@@ -35,6 +35,14 @@ Error SystemNative_CreateNetworkChangeListenerSocket(int32_t* retSocket)
...
     sa.nl_groups = RTMGRP_LINK | RTMGRP_IPV4_IFADDR | RTMGRP_IPV4_ROUTE | RTMGRP_IPV6_ROUTE;
     int32_t sock = socket(AF_NETLINK, SOCK_RAW, NETLINK_ROUTE);
+	
+	// Added receive timeout to prevent recvmsg method in SystemNative_ReadEvents to block thread indefinitely.
+	// This allows method SystemNative_ReadEvents to periodically exit and allows .NET thread to exit if needed.
+	struct timeval tv;
+	tv.tv_sec = 1; //timeout in seconds
+	tv.tv_usec = 0;
+	setsockopt(sock, SOL_SOCKET, SO_RCVTIMEO, (const char*)&tv, sizeof tv);
+
```
Comment (ManickaP) [inline src/native/libs/System.Native/pal_networkchange.c:45]: tabs --> spaces
```suggestion

    // Added receive timeout to prevent recvmsg method in SystemNative_ReadEvents to block thread indefinitely.
    // This allows method SystemNative_ReadEvents to periodically exit and allows .NET thread to exit if needed.
    struct timeval tv;
    tv.tv_sec = 1; //timeout in seconds
    tv.tv_usec = 0;
    setsockopt(sock, SOL_SOCKET, SO_RCVTIMEO, (const char*)&tv, sizeof tv);

```

Comment (ManickaP) [reply src/native/libs/System.Native/pal_networkchange.c:45]: Also this should probably go after the return file descriptor check on lines 49-53, i.e. somewhere after:
```C
    if (sock == -1)
    {
        *retSocket = -1;
        return (Error)(SystemNative_ConvertErrorPlatformToPal(errno));
    }
```

And if the `setsockopt` is necessary for this to work, there should probably be return value check similar to what we have with `bind`.

File: src/libraries/System.Net.NetworkInformation/tests/FunctionalTests/NetworkAddressChangedTests.cs [hunk 2]
```diff
@@ -23,5 +26,56 @@ public void NetworkAddressChanged_JustRemove_Success()
...
         }
+
+        [PlatformSpecific(TestPlatforms.Linux)]
+        [Fact]
+        public async void NetworkAddressChanged_AddRemoveMultipleTimes_CheckForLeakingThreads()
+        {
+            static async Task<int> GetNumberOfNetworkAddressChangeThreadsAsync()
+            {
+                int pid = Process.GetCurrentProcess().Id;
+                ProcessStartInfo psi = new ProcessStartInfo("ps", $"-T -p {pid}");
```
Comment (ManickaP) [inline src/libraries/System.Net.NetworkInformation/tests/FunctionalTests/NetworkAddressChangedTests.cs:37]: Unfortunately calling `ps` fails on some of our docker images:
```
System.Net.NetworkInformation.Tests.NetworkChangedTests.NetworkAddressChanged_AddRemoveMultipleTimes_CheckForLeakingThreads [FAIL]
      System.ComponentModel.Win32Exception : An error occurred trying to start process 'ps' with working directory '/root/helix/work/workitem/e'. No such file or directory
```
I need to follow up with this with our infra, whether to condition the test or fix the images...

**Review by ManickaP — COMMENTED** [review]
Thanks for the contribution!!!
Some comments for the native part.
@wfurt could you also check this out? I don't see anything wrong with this, but I'm also not fully confident I'm not missing something.

File: src/libraries/System.Net.NetworkInformation/tests/FunctionalTests/NetworkAddressChangedTests.cs [hunk 3]
```diff
@@ -23,5 +26,56 @@ public void NetworkAddressChanged_JustRemove_Success()
         {
             NetworkChange.NetworkAddressChanged -= _addressHandler;
         }
+
+        [PlatformSpecific(TestPlatforms.Linux)]
+        [Fact]
+        public async void NetworkAddressChanged_AddRemoveMultipleTimes_CheckForLeakingThreads()
```
Comment (ManickaP) [inline src/libraries/System.Net.NetworkInformation/tests/FunctionalTests/NetworkAddressChangedTests.cs:32]: And we should probably put this into outerloop since it has a delay, but that can be done once we figure out the failing `ps`.

File: src/native/libs/System.Native/pal_networkchange.c (hunk 1 above)
Comment (vdjuric) [reply src/native/libs/System.Native/pal_networkchange.c]: You are right, I will add this after socket was successfully created and then also check setsockopt result.

File: src/libraries/System.Net.NetworkInformation/tests/FunctionalTests/NetworkAddressChangedTests.cs (hunk 2 above)
Comment (vdjuric) [reply src/libraries/System.Net.NetworkInformation/tests/FunctionalTests/NetworkAddressChangedTests.cs:37]: Yes, I noticed few tests were failing because "ps" was missing. I don't know if it is good practice to call external programs from tests. Unfortunately I don't think such test can be done with C# managed code.

**Comment by vdjuric** [comment]
@ManickaP thank you for your suggestions and help!

Created two commits: https://github.com/dotnet/runtime/pull/63963/commits/afd9caaaf7a696521a97a4d545835fdeadcf9c59 and https://github.com/dotnet/runtime/pull/63963/commits/69f536f1eda7e491c6fbf943783d9a1d8b1ace97

I also added TODO comment into pal_networkchange.c to discuss if current one second timeout is good default value for setsockopt, because this means C# thread will call Interop.Sys.ReadEvents every second ...

Also, there is simple workaround for all these issues: on application startup simply register one NetworkChange.NetworkAddressChanged event and don't unsubscribe from it :)

File: src/libraries/System.Net.NetworkInformation/tests/FunctionalTests/NetworkAddressChangedTests.cs (hunk 2 above)
Comment (ManickaP) [reply src/libraries/System.Net.NetworkInformation/tests/FunctionalTests/NetworkAddressChangedTests.cs:37]: We do have some precedent for calling external process from a test and I don't think there's another (managed) way to count the threads by their names.
I found an example of a check that tests a similar thing:
https://github.com/dotnet/runtime/blob/5a12420546d7f0d963bcccf082400c1b057e38ef/src/libraries/System.Net.Security/tests/FunctionalTests/ServerAllowNoEncryptionTest.cs#L74-L74
And the implementation of the check leads to this:
https://github.com/dotnet/runtime/blob/5a9e584d684d45ce67b6e9c9cf27114db20efd1b/src/libraries/System.Net.Security/tests/FunctionalTests/TestConfiguration.cs#L36

Basically trying to call the required CLI command and if it fails disabling the tests that depends on it.

Do you think you could add something like this for this test?

Comment (ManickaP) [reply src/libraries/System.Net.NetworkInformation/tests/FunctionalTests/NetworkAddressChangedTests.cs:37]: For reference, issue to add `ps` into the docker image: https://github.com/dotnet/dotnet-buildtools-prereqs-docker/issues/565

File: src/libraries/System.Net.NetworkInformation/src/System/Net/NetworkInformation/NetworkAddressChange.Unix.cs [hunk 4]
```diff
@@ -16,6 +16,12 @@ public partial class NetworkChange
         // Lock controlling access to delegate subscriptions, socket initialization, availability-changed state and timer.
         private static readonly object s_gate = new object();
 
+        // s_socketCreateContext variable was introduced to prevent event thread leaks on Linux
+        // as described on GitHub: https://github.com/dotnet/runtime/issues/63788
+        // For each CreateSocket() call we increment this variable and pass it to ".NET Network Address Change" thread.
+        // This is then used in method LoopReadSocket as additional condition to exit thread.
+        private static volatile int s_socketCreateContext;
```
Comment (tmds) [inline src/libraries/System.Net.NetworkInformation/src/System/Net/NetworkInformation/NetworkAddressChange.Unix.cs:23]: suggestion: instead of adding `s_socketCreateContext` you can put the socket handle in a reference type.

```cs
class ListenerSocket { public ListenerSocket(int handle); public int Handle; }
static volatile ListenerSocket? s_socket;
```

Comment (vdjuric) [reply src/libraries/System.Net.NetworkInformation/src/System/Net/NetworkInformation/NetworkAddressChange.Unix.cs]: Hi,

Thank you for this suggestion! Seems more elegant solution, I changed this with following commit: https://github.com/dotnet/runtime/pull/63963/commits/e130fdcfa4eb9ce6552d7c9278cd92aaf2a5f383

Comment (tmds) [reply src/libraries/System.Net.NetworkInformation/src/System/Net/NetworkInformation/NetworkAddressChange.Unix.cs]: Looks good.
You can remove `SocketWrapper.NotSet` and use `null` instead.

File: src/libraries/System.Net.NetworkInformation/tests/FunctionalTests/NetworkAddressChangedTests.cs (hunk 2 above)
Comment (vdjuric) [reply src/libraries/System.Net.NetworkInformation/tests/FunctionalTests/NetworkAddressChangedTests.cs]: Hi,

Thank you for this. I think this should now be implemented with https://github.com/dotnet/runtime/pull/63963/commits/d89300329c992c981bd984fa35e8915767f57673 and https://github.com/dotnet/runtime/pull/63963/commits/e94ab24bb170cacd80f518d23049bd7999b75bc2

Also, [OuterLoop()] should be probably enabled for this test as you suggested and remove TODO.

File: src/libraries/System.Net.NetworkInformation/src/System/Net/NetworkInformation/NetworkAddressChange.Unix.cs (hunk 4 above)
Comment (vdjuric) [reply src/libraries/System.Net.NetworkInformation/src/System/Net/NetworkInformation/NetworkAddressChange.Unix.cs]: Initially I had nullable implementation, but then changed it to non-nullable one because I don't like nulls :)

Don't know what is better here ... 

:)

Comment (tmds) [reply src/libraries/System.Net.NetworkInformation/src/System/Net/NetworkInformation/NetworkAddressChange.Unix.cs]: `null` is better. It is a compile-time constant. It's also the default value for fields, and it's a good value to indicate _not set_.

File: src/libraries/System.Net.NetworkInformation/src/System/Net/NetworkInformation/NetworkAddressChange.Unix.cs [hunk 5]
```diff
@@ -149,32 +148,34 @@ private static unsafe void CreateSocket()
...
 
-        private static unsafe void LoopReadSocket(int socket)
+        private static unsafe void LoopReadSocket(SocketWrapper initiallyCreatedSocket)
         {
-            while (socket == s_socket)
+            while (s_socket == initiallyCreatedSocket) //if both references are equal then s_socket was not changed
             {
-                Interop.Sys.ReadEvents(socket, &ProcessEvent);
+                //we can continue processing events
+                Interop.Sys.ReadEvents(initiallyCreatedSocket.Socket, &ProcessEvent);
```
Comment (tmds) [inline src/libraries/System.Net.NetworkInformation/src/System/Net/NetworkInformation/NetworkAddressChange.Unix.cs:178]: Between the condition-check in the while loop, and passing the handle to `ReadEvents` the handle may be re-used for something other. That means we'll be reading for some one else's file descriptor.

Using `Socket`/`SafeSocketHandle` allows to fix that.

File: src/native/libs/System.Native/pal_networkchange.c [hunk 6]
```diff
@@ -44,6 +45,20 @@ Error SystemNative_CreateNetworkChangeListenerSocket(int32_t* retSocket)
         return (Error)(SystemNative_ConvertErrorPlatformToPal(errno));
     }
 
+    // Added receive timeout to prevent recvmsg method in SystemNative_ReadEvents to block thread indefinitely.
```
Comment (tmds) [inline src/native/libs/System.Native/pal_networkchange.c:48]: If we're using `Socket` this is not needed either.

File: src/libraries/System.Net.NetworkInformation/src/System/Net/NetworkInformation/NetworkAddressChange.Unix.cs [hunk 7]
```diff
@@ -149,32 +148,34 @@ private static unsafe void CreateSocket()
                 throw new NetworkInformationException(message);
             }
 
-            s_socket = newSocket;
-            new Thread(s => LoopReadSocket((int)s!))
+            s_socket = new SocketWrapper(newSocket);
+
+            new Thread(args => LoopReadSocket((SocketWrapper)args!))
```
Comment (tmds) [inline src/libraries/System.Net.NetworkInformation/src/System/Net/NetworkInformation/NetworkAddressChange.Unix.cs:153]: and using `Socket` we could do async reading, instead of making a thread.

**Comment by tmds** [comment]
I've highlighted what we can improve further by using `Socket`.
@vdjuric do you want to look into it? Or otherwise, we can tackle it in a follow-up PR.

**Comment by vdjuric** [comment]
@tmds I think those Socket improvements would be better as part of another PR. Regarding async Socket reading I also think this would be better than creating new Thread.

Using null instead of SocketWrapper.NotSet can be easily rewritten, this is not a problem.

I am wondering why some tests are failing although conditional fact was added. One of the logs reports "Exception while trying to run 'ps' command for user 'helixbot'" (I added additional checkin for diagnostics to report which user is running this), another returns "Process 'ps' returned exit code 1".

Those exceptions should also happen during conditional fact check, which should return "false" --> test should be skipped.

```c#
if (OperatingSystem.IsWindows())
{
	return false;
}

// On other platforms we will try it.
try
{
	_ = ProcessUtil.GetProcessThreadsWithPsCommand(Process.GetCurrentProcess().Id);
	return true;
}
catch { return false; }
```

**Comment by ManickaP** [comment]
The static method that is evaluated in the `ConditionalFact` mas be accessible within the test class. I.e.: you need to add simple static property to `NetworkChangedTests` like this:
https://github.com/dotnet/runtime/blob/5a12420546d7f0d963bcccf082400c1b057e38ef/src/libraries/System.Net.Security/tests/FunctionalTests/ServerAllowNoEncryptionTest.cs#L97

**Comment by vdjuric** [comment]
Hi,

I think this is added:

```C#
        [PlatformSpecific(TestPlatforms.Linux)]
        //[OuterLoop()] //TODO: add Outer Loop attribute?
        [ConditionalFact(nameof(SupportsGettingThreadsWithPsCommand))]
        public void NetworkAddressChanged_AddRemoveMultipleTimes_CheckForLeakingThreads()
        {
            for (int i = 1; i <= 10; i++)
            {
                NetworkChange.NetworkAddressChanged += _addressHandler;
                NetworkChange.NetworkAddressChanged -= _addressHandler;
            }

            Thread.Sleep(2000); //allow some time for threads to exit

            //We are searching for threads containing ".NET Network Ad"
            //because ps command trims actual thread name ".NET Network Address Change".
            //This thread is created in:
            //  src/libraries/System.Net.NetworkInformation/src/System/Net/NetworkInformation/NetworkAddressChange.Unix.cs
            int numberOfNetworkAddressChangeThreads = ProcessUtil.GetProcessThreadsWithPsCommand(Process.GetCurrentProcess().Id)
                .Where(e => e.IndexOf(".NET Network Ad") > 0).Count();

            Assert.Equal(0, numberOfNetworkAddressChangeThreads); //there should be no threads because there are no event subscribers
        }

        private static bool SupportsGettingThreadsWithPsCommand
            => TestConfiguration.SupportsGettingThreadsWithPsCommand;
```

For local test I also hardcoded property SupportsGettingThreadsWithPsCommand to always return false and it was skipped from local tests.

**Comment by ManickaP** [comment]
As to the socket improvements, the change in this PR as-is is not making the existing situation worse so we should proceed with getting this in. The suggested changes can be tackled separately. @tmds can you file an issue for them or should I?

**Comment by ManickaP** [comment]
Got it, the function that calls the process uses `yield return`. Since you throw away the resulting `IEnumerable` and never enumerate the iterator, it will never get invoked, thus won't throw in the test for `ConditionalFact`.
You can just add simple `ToList()`, `FirstOrDefault()` etc. or rewrite the function to return pre-calculated array/list and it'll start working as expected.

**Comment by vdjuric** [comment]
Wow, very very complex code :) Sorry for this! In https://github.com/dotnet/runtime/pull/63963/commits/821a5d13db8546e952199a3c26cc0b339f7bb147 this is now replaced with more simple version (`List<string>`).

It seems I don't have luck, now those few tests are canceled. Linux doesn't like me :)

**Comment by vdjuric** [comment]
Please ignore my previous post, I think I understand now, will do another experimental checkin ...

File: src/libraries/System.Net.NetworkInformation/src/System/Net/NetworkInformation/NetworkAddressChange.Unix.cs [hunk 8]
```diff
@@ -290,5 +291,21 @@ private static void OnAvailabilityTimerFired(object? state)
...
+        private class SocketWrapper
+        {
+            public SocketWrapper(int socket)
+            {
+                Socket = socket;
+            }
+
+            public int Socket { get; }
+
+            public static readonly SocketWrapper NotSet = new SocketWrapper(0);
```
Comment (wfurt) [inline src/libraries/System.Net.NetworkInformation/src/System/Net/NetworkInformation/NetworkAddressChange.Unix.cs:308]: While unlikely, 0 is valid handle. We should perhaps use -1

File: src/libraries/System.Net.NetworkInformation/src/System/Net/NetworkInformation/NetworkAddressChange.Unix.cs (hunk 5 above)
Comment (wfurt) [reply src/libraries/System.Net.NetworkInformation/src/System/Net/NetworkInformation/NetworkAddressChange.Unix.cs:178]: That may be bigger change than  @vdjuric may want to take.  If we feel that would simplify the code, I think we should open separate issue for it @tmds. 
Alternatively, we can possibly use `SafeHandle` but that would also be much bigger change IMHO.

File: src/libraries/System.Net.NetworkInformation/src/System/Net/NetworkInformation/NetworkAddressChange.Unix.cs (hunk 8 above)
Comment (vdjuric) [reply src/libraries/System.Net.NetworkInformation/src/System/Net/NetworkInformation/NetworkAddressChange.Unix.cs]: Hi, thank you for this!

With commit https://github.com/dotnet/runtime/pull/63963/commits/b6e944d44b40ec6e68b3b69b73021fd538558ffd this is now changed, we simply set `SocketWrapper` to null as suggested by @tmds

**Comment by vdjuric** [comment]
Mini dilemmas for https://github.com/dotnet/runtime/pull/63963/commits/b6e944d44b40ec6e68b3b69b73021fd538558ffd in method `CloseSocket`:

```C#
        private static void CloseSocket()
        {
            Debug.Assert(s_socket != null, "s_socket was null when CloseSocket was called.");
            Interop.Error result = Interop.Sys.CloseNetworkChangeListenerSocket(s_socket != null ? s_socket.Socket : -1);
            if (result != Interop.Error.SUCCESS)
            {
                string message = Interop.Sys.GetLastErrorInfo().GetErrorMessage();
                throw new NetworkInformationException(message);
            }

            s_socket = null;
        }
```

If `s_socket` would be null when calling this method, should we:

1. leave code as
2. explicitly throw exception instead of line with Debug.Assert
2. quietly ignore this after Debug.Assert and simply return from function before `Interop.Sys.CloseNetworkChangeListenerSocket` is called (no exceptions are thrown)

**Comment by tmds** [comment]
> leave code as is (we will pass -1 to Interop.Sys.CloseNetworkChangeListenerSocket and this should return some error)

Yes, this can stay as is.

File: src/libraries/System.Net.NetworkInformation/tests/FunctionalTests/NetworkAddressChangedTests.cs [hunk 9]
```diff
@@ -23,5 +26,31 @@ public void NetworkAddressChanged_JustRemove_Success()
         {
             NetworkChange.NetworkAddressChanged -= _addressHandler;
         }
+
+        [PlatformSpecific(TestPlatforms.Linux)]
+        //[OuterLoop()] //TODO: add Outer Loop attribute?
```
Comment (ManickaP) [inline src/libraries/System.Net.NetworkInformation/tests/FunctionalTests/NetworkAddressChangedTests.cs:31]:
```suggestion
        [OuterLoop("May take several seconds")]
```

**Comment by ManickaP** [comment]
Interop failures unrelated: https://github.com/dotnet/runtime/issues/64172

**Review by ManickaP — APPROVED** [review]
Failures are unrelated, added test runs as expected.
And I think all outstanding comments are resolved or shelved, so we can merge this.

**Comment by ManickaP** [comment]
Thanks @vdjuric for the contribution and fixing this!

**Comment by vdjuric** [comment]
Hi, thank you very much and no problem at all, you all helped me very much! It was great learning experience (first time building .NET with WSL :) ).
